TrafficCapture/captureKafkaOffloader/src/main/java/org/opensearch/migrations/trafficcapture/kafkaoffloader/KafkaCaptureFactory.java

```diff
@@ -104,62 +96,47 @@ public CodedOutputStreamWrapper createStream() {
-
-                // Async request to Kafka cluster
-                producer.send(kafkaRecord, handleProducerRecordSent(cf, recordId, flushContext));
-                metricsLogger.atSuccess(MetricsEvent.RECORD_SENT_TO_KAFKA)
-                        .setAttribute(MetricsAttributeKey.CHANNEL_ID, connectionId)
-                        .setAttribute(MetricsAttributeKey.TOPIC_NAME, topicNameForTraffic)
-                        .setAttribute(MetricsAttributeKey.SIZE_IN_BYTES, kafkaRecord.value().length)
-                        .setAttribute(MetricsAttributeKey.REQUEST_ID, recordId).emit();
-                return cf;
-            } catch (Exception e) {
-                metricsLogger.atError(MetricsEvent.RECORD_FAILED_TO_KAFKA, e)
-                        .setAttribute(MetricsAttributeKey.CHANNEL_ID, connectionId)
-                        .setAttribute(MetricsAttributeKey.TOPIC_NAME, topicNameForTraffic).emit();
-                throw e;
-            }
+
+            String recordId = String.format("%s.%d", connectionId, index);
+            var byteBuffer = osh.byteBuffer;
+            ProducerRecord<String, byte[]> kafkaRecord = new ProducerRecord<>(topicNameForTraffic, recordId,
+                    Arrays.copyOfRange(byteBuffer.array(), 0, byteBuffer.position()));
+            // Used to essentially wrap Future returned by Producer to CompletableFuture
+            var cf = new CompletableFuture<RecordMetadata>();
+            log.debug("Sending Kafka producer record: {} for topic: {}", recordId, topicNameForTraffic);
+
+            var flushContext = rootScope.createKafkaRecordContext(telemetryContext,
+                    topicNameForTraffic, recordId, kafkaRecord.value().length);
+
+            // Producer Send will block on calls such as retrieving cluster metadata, running fully async
+            return CompletableFuture.supplyAsync(() -> {
+                    try {
+                        /*
+                         * The default KafkaProducer comes with built-in retry and error-handling logic that suits many cases. From the
+                         * documentation here for retry: https://kafka.apache.org/35/javadoc/org/apache/kafka/clients/producer/KafkaProducer.html
+                         * "If the request fails, the producer can automatically retry. The retries setting defaults to Integer.MAX_VALUE,
+                         * and it's recommended to use delivery.timeout.ms to control retry behavior, instead of retries."
+                         *
+                         * Apart from this the KafkaProducer has logic for deciding whether an error is transient and should be
+                         * retried or not retried at all: https://kafka.apache.org/35/javadoc/org/apache/kafka/common/errors/RetriableException.html
+                         * as well as basic retry backoff
+                         */
+                        return producer.send(kafkaRecord).get();
```

Comment:
If we can take advantage of the callback notifying that we're done, we can use that as an opportunity to provide the signal to know when we can/should get the RecordMetadata.  Here are two alternate ways that this can be done.  I prefer the second one because there are less CompletableFuture objects being constructed (and they can be miserable to figure out why work hasn't proceeded).

What do you think?

```
        private CompletableFuture<RecordMetadata> safeSendA(KafkaRecordContext flushContext, 
                                                          ProducerRecord<String, byte[]> kafkaRecord, String recordId) {
            // producer.send() will block on some pre-requisite calls such as retrieving cluster metadata,
            // running call immediately from an async keeps this code fully non-blocking
            CompletableFuture<Void> kafkaCallbackDone = new CompletableFuture<>();
            return CompletableFuture.supplyAsync(() -> {
                        log.debug("Sending Kafka producer record: {} for topic: {}", recordId, topicNameForTraffic);
                        try {
                            return producer.send(kafkaRecord, (x, y) -> kafkaCallbackDone.complete(null));
                        } catch (Exception e) {
                            kafkaCallbackDone.completeExceptionally(e);
                            throw Lombok.sneakyThrow(e);
                        }
                    })
                    .thenCombine(kafkaCallbackDone, (f,v)-> {
                        try {
                            return f.get();
                        } catch (InterruptedException e) {
                            Thread.currentThread().interrupt();
                            throw Lombok.sneakyThrow(e);
                        } catch (ExecutionException e) {
                            throw Lombok.sneakyThrow(e);
                        }
                    })
                    .whenComplete(((recordMetadata, throwable) -> {
                        if (throwable != null) {
                            flushContext.addException(throwable);
                            log.error("Error sending producer record: {}", recordId, throwable);
                        } else {
                            log.debug("Kafka producer record: {} has finished sending for topic: {} and partition {}",
                                    recordId, recordMetadata.topic(), recordMetadata.partition());
                        }
                        flushContext.close();
                    }));
        }


        private CompletableFuture<RecordMetadata> safeSendB(KafkaRecordContext flushContext,
                                                          ProducerRecord<String, byte[]> kafkaRecord, String recordId) {
            // producer.send() will block on some pre-requisite calls such as retrieving cluster metadata,
            // running call immediately from an async keeps this code fully non-blocking
            CompletableFuture<RecordMetadata> kafkaCallbackDone = new CompletableFuture<>();
            ForkJoinPool.commonPool().execute(() -> {
                log.debug("Sending Kafka producer record: {} for topic: {}", recordId, topicNameForTraffic);
                try {
                    producer.send(kafkaRecord, (recordMetadata, exception) -> {
                        if (exception == null) {
                            kafkaCallbackDone.complete(recordMetadata);
                        } else {
                            kafkaCallbackDone.completeExceptionally(exception);
                        }
                    });
                } catch (Exception e) {
                    kafkaCallbackDone.completeExceptionally(e);
                    throw Lombok.sneakyThrow(e);
                }
            });
            kafkaCallbackDone.whenComplete(((recordMetadata, throwable) -> {
                if (throwable != null) {
                    flushContext.addException(throwable);
                    log.error("Error sending producer record: {}", recordId, throwable);
                } else {
                    log.debug("Kafka producer record: {} has finished sending for topic: {} and partition {}",
                            recordId, recordMetadata.topic(), recordMetadata.partition());
                }
                flushContext.close();
            }));
            return kafkaCallbackDone;
        }
```